### PR TITLE
[CFP-158] Fix govuk_design_system_formbuilder nested error linking

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
Fix govuk_design_system_formbuilder nested error linking

#### Ticket

[CFP-156](https://dsdmoj.atlassian.net/browse/CFP-156)

#### Why
error linking between summary and field are broken nested fields when using the govuk formbuilder

#### How
- [X] Name nested errors following expectations from form builder
- [X] Use concern to override nested error attribute naming for validators on models using formbuilder on FE
- [X] Use DfE formbuilder 2.7.2

#### TODO
- [x] use new version of govuk formbuilder once [fix-multiple-captures](https://github.com/DFE-Digital/govuk_design_system_formbuilder/pull/301) is merged and released
- [x] unit spec error naming via composition and inheritance
- [x] feature spec unhappy path for provider supplier number error linking?!

### CODE REVIEW

File: app/views/shared/_supplier_numbers.html.haml
Comment (naseberry) [inline app/views/shared/_supplier_numbers.html.haml:4]: 👍

File: config/locales/en.yml
Comment (naseberry) [inline config/locales/en.yml:222]: Well spotted.

File: features/provider_management.feature
Comment (naseberry) [inline features/provider_management.feature]: Are these comments **to be or not to be**?

File: features/step_definitions/govuk_component_steps.rb
Comment (naseberry) [inline features/step_definitions/govuk_component_steps.rb:1]: Would it be ideal to extend this new step definition  to include govuk textbox and other govuk form inputs?

**Comment by naseberry** [comment]
I've manually tested the provider views on the dev-lgfs environment, error linking now works as intended with  nested forms.

I would suggest we get a second review from a BE developer.

👍 Great work.

PS - I noticed the error summary order was incorrect, will this be in a separate PR?

File: features/provider_management.feature
Comment (jsugarman) [reply features/provider_management.feature]: yes, this is the question of preference i wanted to raise in the commits comment with you guys
> I prefer the more succinct cucumber table as a reusable
pattern for future unhappy path scenarios. However,
this could be more confusing during failures so would
like other devs views and preferences input.

What do you think?

File: features/step_definitions/govuk_component_steps.rb
Comment (jsugarman) [reply features/step_definitions/govuk_component_steps.rb:1]: I think so, I added [this commit](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/4080/commits/701de5ef718c1f931293e17df210e950a2bbb79e) to move other steps I found to this file. Are there others you are thinking-of/have-found?

**Comment by jsugarman** [comment]
Cheers
> PS - I noticed the error summary order was incorrect, will this be in a separate PR?

Error order is not controlled in the provider form and falls back to default rails/activemodel behaviour I believe. 

As part of the spike I used to test custom error handling I believe it would work if providers were validated using our "custom error handling", but they currently do not. I have not tried to implement the custom error handling on this form as yet as yak shaving will undoubtedly ensue. 

I think we could apply custom error handling to this form as part of the ticket to handle formbuilder working with custom error handling as 1. a prelude to applying formbuilder to claim forms 2. to standardise our error handling further across different objects (we currently use either the rails-way or our-way)

File: spec/validators/claim/base_claim_sub_model_validator_spec.rb
Comment (jrmhaig) [inline spec/validators/claim/base_claim_sub_model_validator_spec.rb]: Does this need to be a `subject`? I don't think it is being used as the subject of any tests. Likewise for the other places where the new shared examples are used.

Not a major issue. Maybe it is clearer like this.

File: features/provider_management.feature
Comment (jrmhaig) [reply features/provider_management.feature]: The tabular form is more understandable. What happens for a failure? Does it just indicate a failure without specifying which line? I don't think that is a problem, especially if there are just a few lines.

Comment (jsugarman) [reply features/provider_management.feature]: its not too bad, it will fail on the first row and expectation but from my testing it is pretty easy to identify the failure

File: spec/validators/claim/base_claim_sub_model_validator_spec.rb
Comment (jsugarman) [reply spec/validators/claim/base_claim_sub_model_validator_spec.rb]: have added [this commit](https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/pull/4080/commits/db6be30299f56577ee458f158b4c7b0a315b56e4)  which makes it more consistent and places the subject in its "proper" context  - i.e. the class under test.

**Review by jrmhaig — APPROVED** [review]
:+1: Looks OK.

File: features/provider_management.feature
Comment (naseberry) [reply features/provider_management.feature]: ^^ I agree with Joe, the tabular form is understandable for me

File: features/step_definitions/govuk_component_steps.rb
Comment (naseberry) [reply features/step_definitions/govuk_component_steps.rb:1]: I was thinking more of https://github.com/ministryofjustice/Claim-for-Crown-Court-Defence/blob/master/features/step_definitions/common_steps.rb#L204

Adding specificity like the other govuk steps by finding the relevant class name, but then again, I guess the reference above seem loose and robust.
What do you think?

**Pull request merged**
